### PR TITLE
Tutorial registry

### PR DESCRIPTION
**Pull request by grst**
The idea here is to create a tutorial registry that works analogous to the [ecosystem package registry](https://github.com/scverse/ecosystem-packages/). For each tutorial, a `meta.yml` file would be added that includes a link to the tutorial, a preview image and some metadata which is validated against a `json-schema`. 

The list of tutorials could include both core tutorials (such those developed at the hackathon), core-package tutorials, and external tutorials. External (ecosystem package) tutorials could be added by their developers via a PR if certain criteria are met (see README in this PR, copied from https://github.com/scverse/governance/pull/44). 

For rendering the tutorial overview page, I envisage a page similar to what is currently at `scverse.org/learn`. It would be nice if the metadata showed as tags (e.g. difficulty  / modality / category) and if there was some way of filtering the tutorials (similar to the rendering of the ecosystem packages). 

### CI checks
 - [ ] json schema
 - [ ] links are point to existing websites
 - [ ] an image with dimensions 512x512px (or whatever) is available for each tutorial.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -8,5 +8,44 @@ The tutorials found here are _shared tutorials_ that complement more specific tu
...
 
+# tutorial registry
+
+The purpose of this document is to formalize the criteria to be eligible for external tutorial inclusion into the scverse tutorials listing (https://scverse.org/scverse-tutorials/).
+
+## Criteria to be eligible for tutorial inclusion
+
+We have the following requirement for a package to be eligible to be considered for tutorial submission/mirroring:
+
+1. The package must be a scverse listed ecosystem package and therefore meet all corresponding requirements.
```
Comment (Zethson) [inline README.md:19]: Can we link to the requirements/process here?

File: README.md [hunk 2]
```diff
@@ -8,5 +8,44 @@ The tutorials found here are _shared tutorials_ that complement more specific tu
...
+2. Any used (advanced) methods must be published. Preprints are acceptable on a case by case basis.
```
Comment (Zethson) [inline README.md:20]: It's not quite clear what "on a case by case basis" here means. Is it us judging this? Should we ask people with preprints to contact us beforehand ?

File: README.md [hunk 3]
```diff
@@ -8,5 +8,44 @@ The tutorials found here are _shared tutorials_ that complement more specific tu
...
+3. We require an active point of contact that is available on zulip. We reserve the right to remove packages if we cannot reach a responsible person anymore.
```
Comment (Zethson) [inline README.md:21]:
```suggestion
3. We require an active point of contact that is available on the [scverse zulip](https://scverse.zulipchat.com/). We reserve the right to remove packages if we cannot reach a responsible person anymore.
```

File: tutorial-registry/schema.json [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            "description": "Summary of the tutorial (1-2 sentences).",
+            "type": "string"
+        },
+        "link": {
+            "description": "Link to the tutorial",
+            "type": "string",
+            "format": "uri"
+        },
+        "level": {
+            "description": "Difficulty of the tutorial",
```
Comment (Zethson) [inline tutorial-registry/schema.json:22]: Maybe two categories are enough? Not sure how important this is to begin with. Ideally they're all simple and structured.

File: tutorial-registry/schema.json [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            "type": "string",
+            "enum": ["beginner", "intermediate", "advanced"]
+        },
+        "category": {
+            "description": "Category of the tutorial",
```
Comment (Zethson) [inline tutorial-registry/schema.json:27]: Really tricky to nail the categories with the right resolution. Maybe add

`machine learning`

Why is `data integration` its own category?

File: tutorial-registry/schema.json [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+                "analysis",
+                "data integration",
+                "interoperability",
+                "visualization"
+            ]
+        },
+        "modality": {
+            "description": "Is the tutorial related to a specific modality?",
+            "type": "string",
+            "enum": [
```
Comment (Zethson) [inline tutorial-registry/schema.json:40]: Maybe order it slightly differently: DNA before RNA. Spatial before multimodal

File: tutorial-registry/schema.json [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        },
+        "packages": {
+            "type": "array",
+            "items": {
+                "type": "string"
+            },
+            "description": "List the scverse core and ecosystem packages this tutorial is about. "
+        },
+        "license": {
+            "description": "The license of the tutorial",
```
Comment (Zethson) [inline tutorial-registry/schema.json:59]: Should we offer options here? Or even enforce a license? I doubt that we want 10 different licenses.

File: tutorial-registry/schema.json (hunk 4 above)
Comment (grst) [reply tutorial-registry/schema.json:22]: It's more about where to start first. E.g. last hackathon we discussed you should have understood single-cell processing using `normalize_total` + `log1p` before moving to processing with scvi-tools. 

Also you should have understood what the AnnData slots are for before reading a tutorial how to use distributed dask arrays with AnnData.

File: tutorial-registry/schema.json (hunk 5 above)
Comment (grst) [reply tutorial-registry/schema.json:27]: agreed. I don't think this list is final, were just a few terms that came to my mind.

File: tutorial-registry/schema.json (hunk 7 above)
Comment (grst) [reply tutorial-registry/schema.json:59]: it should probably some sort of CC license...

File: tutorial-registry/schema.json (hunk 5 above)
Comment (ivirshup) [reply tutorial-registry/schema.json:27]: Do we need to start out with an enum here? I feel like we'd end up modifying the schema quite often

Comment (grst) [reply tutorial-registry/schema.json:27]: I feel if we don't we end up with one category per tutorial :shrug: 
I also don't think it's a problem to add an additional category to the schema in the same PR adding a new tutorial using that category.

**Review by ivirshup — COMMENTED** [review]
Thanks for doing this!

## scope?

What's the scope of this? To me, the learn page and this tutorial package have a pretty curated small set of tutorials. How many things do you think will go here? 

Is this supposed to be the one location to go to for scverse learning material, or just a place? I had thought more the latter.

## Schema

I think the schema will be quite useful for making the site. 

## Requirements

I'm not a huge fan of the restrictions, especially the detail. I think it's fine to say we have broad discretion and leave the more specific criteria to a community effort or something down the road.

This has a lot to do with the scope. I don't really want to be in the position of gatekeeping learning materials. I think that is more the job of something like sc-best-practices.org. I would like to make it easier to learn how to use scverse tools.

File: README.md (hunk 1 above)
Comment (grst) [reply README.md:19]: I guess all of them should be core or listed ecosystem packages?

**Comment by grst** [comment]
> scope

I think the learn page should provide an overview of all sorts of analyses you can do with scverse, not just core packages. It should complement the ecosystem packages page. If you you are seeking to perform a certain type of analysis, I think it's quite hard to find the proper package from the ecosystem page unless you already know what to look for, while with tutorials this should be easier (For instance, if you want to do pseudobulk DE, there's no way to know you need the `decoupler` package for that). 

But I agree it shouldn't include *all* ecosystem package tutorials. Just as an example from scirpy, I would include the [analysis tutorial](https://scverse.org/scirpy/latest/tutorials/tutorial_3k_tcr.html) because it showcases a type of analysis (TCR) that can be done with scverse, but not the [data loading tutorial](https://scverse.org/scirpy/latest/tutorials/tutorial_io.html). Similarly, for decoupler, I'd show the [pseudobulk functional analysis](https://decoupler-py.readthedocs.io/en/latest/notebooks/pseudobulk.html) tutorial, because it's another fundamental type of analysis, but not the [cell-type annotation](https://decoupler-py.readthedocs.io/en/latest/notebooks/cell_annotation.html) tutorial, because this is covered by a tutorial on [scverse-tutorials](https://github.com/scverse/scverse-tutorials/pull/21). 

Now how do we formalize this :see_no_evil:? Another option is to keep this vague: 

> On the `scverse.org/learn` page we aim at providing a comprehensive overview of analyses that can be performed with scverse core and ecosystem packages. If you think a tutorial should be linked here, please open an issue. We will discuss the request in the next open community meeting. 

(this could still be followed by a list of rules that are necessary, but not necessarily sufficient for a tutorial to be included)

**Comment by ivirshup** [comment]
> (this could still be followed by a list of rules that are necessary, but not necessarily sufficient for a tutorial to be included)

Which rules would you keep?

**Comment by grst** [comment]
> Which rules would you keep?

I think just 
 * all featured packages must be approved scverse ecosystem packages
 * the notebook author agrees to maintain the tutorial in the future and is reachable via zulip.

**Comment by ivirshup** [comment]
> all featured packages

Broadly agree, but what about packages that aren't domain specific? Like seaborn, xarray, pytorch?

**Comment by grst** [comment]
> Broadly agree, but what about packages that aren't domain specific? Like seaborn, xarray, pytorch?

I specifically meant domain-specific packages, aka packages that interact with AnnData and MuData

**Comment by ivirshup** [comment]
Cool, then I like this approach.

**Comment by grst** [comment]
updated the README accordingly. LMK what you think!

File: tutorial-registry/schema.json (hunk 5 above)
Comment (grst) [reply tutorial-registry/schema.json:27]: ok, let's rename this to "topic" and allow multiple values. 

I would still be in favor of an enum that is adapted dynamically. Without, it's hard to know which topics already exist and similar ones will be created. 

This is what we have now as tags for ecosystem packages, which is pretty much useless: 
```
Counter({'GRN': 1,
         'TF': 3,
         'ML': 1,
         'cell-fate': 2,
         'rna-velocity': 1,
         'trajectory-generation': 2,
         'BLAST': 1,
         'visualization': 1,
         'doublet': 1,
         'single cell': 2,
         'optimal transport': 1,
         'multi omics': 1,
         'data integration': 2,
         'NMF': 1,
         'pathology': 1,
         'rna-seq': 1,
         'differential-expression': 1,
         'SnapATAC': 1,
         'R package': 1,
         'spatial analysis': 1,
         'single-cell': 2,
         'disentanglement': 1,
         'generative framework': 1,
         'Bayesian': 1,
         'HCA': 1,
         'dandelion': 1,
         'bcr': 1,
         'tcr': 1,
         'GSEA': 1,
         'Pathway': 1,
         'vector': 1,
         'scanpy': 1,
         'epigenomics': 1,
         'gene-signatures': 1,
         'CNV': 1,
         'transcriptome analysis': 1,
         'regulatory networks': 1,
         'clustering': 2,
         'scATAC-seq': 1,
         'pseudotime': 1,
         'perturbation': 1,
         'benchmarking': 1,
         'rna velocity': 1,
         'data repository': 1,
         'signals': 1,
         'filter': 1,
         'spatial alignment': 1,
         'spatial registration': 1,
         'spatial decomposition': 1,
         'spatial mapping': 1,
         'imaging': 1})
```

Comment (grst) [reply tutorial-registry/schema.json:27]: My initial idea was to have one "category" per tutorial, where each category has some descriptive text (see `categories.yml`) that is used to group the tutorials on the website, similar to how this is done on the corresponding [Seurat page](https://satijalab.org/seurat/articles/get_started.html).

E.g. "data integration" would be such a category in the following screenshot: 

image

Comment (Zethson) [reply tutorial-registry/schema.json:27]: Yeah, I think that this is great and makes sense

Comment (ivirshup) [reply tutorial-registry/schema.json:27]: > My initial idea was to have one "category" per tutorial, where each category has some descriptive text (see categories.yml) that is used to group the tutorials on the website

That makes sense to me, but then is this the source that generates the learn page? E.g., the `tutorials` yaml entries here: 

https://github.com/scverse/scverse.github.io/blob/06b3b34f3dba223b06eacda57c085b403cb2e016/content/learn/_index.md?plain=1#L11-L20

Comment (grst) [reply tutorial-registry/schema.json:27]: >  but then is this the source that generates the learn page

Yes, that would be the plan. Same principle as for the ecosystem packages page.

**Comment by grst** [comment]
one more requirement: add backlink to `scverse.org/learn`, e.g. using a see also [admonition](https://myst-parser.readthedocs.io/en/latest/syntax/admonitions.html).

File: README.md [hunk 8]
```diff
@@ -2,11 +2,44 @@
...
+-   a registry for tutorials listed on [scverse.org/learn](https://scverse.org/learn) (see `tutorial-registry`)
+-   shared tutorials that complement more specific tutorials provided by invidiual [core](https://scverse.org/packages/)
+    and [ecosystem](https://scverse.org/packages/#ecosystem) packages (see `docs`)
+
+## Adding tutorials
+
+If you believe a tutorial should be added to `scverse.org/learn`, please open an issue. We will discuss the request
+in the next [open community meeting](https://hackmd.io/VfVLKb3ETGKN2j_7tn8ZJQ?view).
+
+For being added to our website, tutorials must fulfil at least the following requirements:
```
Comment (Zethson) [inline README.md:19]:
```suggestion
To be added to our website, tutorials must fulfil at least the following requirements:
```

Comment (Zethson) [inline README.md:19]:
```suggestion
For being added to our website, tutorials must fulfill at least the following requirements:
```

File: README.md [hunk 9]
```diff
@@ -2,11 +2,44 @@
...
+-   the notebook contains a backlink to [scverse.org/learn](scverse.org/learn)
+
+## Structure of external tutorials
+
+While we do not mandate a specific structure for tutorials,
+a good tutorial typically comprises the following sections:
+
+1. **General header**: The tutorial should have a general header that corresponds to the analysis.
+2. **Brief introduction**: The tutorial should introduce the package, the analysis motivation and potentially biological background.
+3. **Requirements to run the notebook**: Special computational requirements like memory or GPUs should be specified. Any required input from other notebooks should also be listed here.
```
Comment (Zethson) [inline README.md:34]: I'd like to mention somewhere that these tutorials should be somewhat self contained and not require weird input from some random places. Not sure whether this is the best place to enforce/mention this.

**Comment by grst** [comment]
Merging, will iterate in follow-up PR

**Pull request merged**